### DISCUSSION TITLE
Building on macOS: Ran into trouble with `iconv`

### DISCUSSION THREAD

**Discussion by danfuzz** [pulsar-edit/.github#225] [Hacking Pulsar]
I'm trying to get a Pulsar build up and running on my macOS laptop, and I've run into a problem which I'm not sure how to address. In particular, I don't know what specifically to blame for the following errors that were produced when I did `yarn install`:

```
gyp info it worked if it ends with ok
gyp info using node-gyp@9.1.0
gyp info using node@16.20.2 | darwin | arm64
gyp info find Python using Python version 3.9.6 found at "/Library/Developer/CommandLineTools/usr/bin/python3"
[...]
gyp info spawn make
gyp info spawn args [ 'BUILDTYPE=Release', '-C', 'build' ]
  CXX(target) Release/obj.target/superstring_core/src/core/encoding-conversion.o
../src/core/encoding-conversion.cc:64:13: error: no matching function for call to 'iconv_close'
  if (data) iconv_close(data);
            ^~~~~~~~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/iconv.h:74:5: note: candidate function not viable: cannot convert argument of incomplete type 'void *' to 'iconv_t' (aka '__tag_iconv_t *') for 1st argument
int     iconv_close(iconv_t);
        ^
../src/core/encoding-conversion.cc:120:32: error: no matching function for call to 'iconv'
      auto conversion_result = iconv(
                               ^~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk/usr/include/iconv.h:71:8: note: candidate function not viable: no known conversion from 'iconv_t *' (aka '__tag_iconv_t **') to 'iconv_t' (aka '__tag_iconv_t *') for 1st argument; dereference the argument with *
size_t  iconv(iconv_t, char ** __restrict,
        ^
2 errors generated.
make: *** [Release/obj.target/superstring_core/src/core/encoding-conversion.o] Error 1
gyp ERR! build error 
```

Near as I can tell, I am performing the build in an environment wherein the expected version of Node is installed. That said, I strayed from the official Pulsar build instructions and used explicit `PATH` manipulation to set it up instead of using `nvm` (which I don't have installed; I just use Homebrew and don't want to mess with my usual setup for this foray):

```
$ export LDFLAGS="-L/opt/homebrew/opt/node@16/lib"
$ export CPPFLAGS="-I/opt/homebrew/opt/node@16/include"
$ PATH="/opt/homebrew/opt/node@16/bin:$PATH"
$ yarn install
[...]
```

While I _believe_ what I did to set up my environment is kosher, I'm not 100% sure. I'd be most appreciative of any guidance. Thanks in advance.

**Comment by Daeraxa** [comment]
> That said, I strayed from the official Pulsar build instructions

In fairness, you didn't :). The official instructions for building on macOS are still somewhat... lacking.

![image](https://github.com/pulsar-edit/.github/assets/58074586/3c13d9eb-50ef-46d7-b661-af356e148beb)

Might be worth checking against the [Cirrus build script](https://github.com/pulsar-edit/pulsar/blob/master/.cirrus.yml#L120) (which does actually use `brew` to manage node).

**Comment by danfuzz** [comment]
Thanks for the lead. I have Xcode 15 installed, and the Cirrus script seems to reference Xcode 14. This seems like a salient difference, though I don't have a particular clue how to proceed. (This isn't my usual area of work. I program on a Mac, but I don't exactly do Mac programming.)

**Comment by confused-Techie** [comment]
@danfuzz One thing I'd take a look at is the requirements of `node-gyp` on macOS, since that's where this is failing, here's their [docs](https://github.com/nodejs/node-gyp#on-macos).

`node-gyp` Is used to build all of our native modules, and as long as your local repo is completely up to date (As of the release just the other day) you should be able to follow their requirements there exactly, since we just recently got up to date there.

Which would mean having:
* Xcode Command Line Tools (Which should install `clang`, `clang++`, and `make`
* They say a supported version of Python but I recommend Python 3.11 (3.12 introduced breaking changes to `node-gyp` again)

Hopefully with those installed you should be good to go

**Comment by danfuzz** [comment]
Unclear if this is actually the problem, but I did find <https://stackoverflow.com/questions/57734434/libiconv-or-iconv-undefined-symbol-on-mac-osx>, which describes a similar situation. The accepted-answer looks like it might be spot-on.

**Comment by danfuzz** [comment]
**Resolved!** (I think)

The SO link (above) helped, but it turned out that a later answer was key. `brew install libiconv`, along with more editing of the flags env vars. With that, I've been able to get through the build without error, and run the result. I won't go so far as to suggest that the following necessarily become the "official" macOS build instructions, but for the record here's what I did:

```bash
$ brew install node@16
$ brew install corepack
$ brew install libiconv
$ export LDFLAGS="-L/opt/homebrew/opt/node@16/lib -L/opt/homebrew/opt/libiconv/lib"
$ export CPPFLAGS="-I/opt/homebrew/opt/node@16/include -I/opt/homebrew/opt/libiconv/include"
$ PATH="/opt/homebrew/opt/node@16/bin:$PATH"
$ corepack enable

$ cd <pulsar source directory>
$ git submodule init && git submodule update
[... and all subsequent steps as listed in the build doc ...]
```

**Reply by confused-Techie** [reply]
Thanks a ton for reporting back what got it working for you.

And very well maybe it should become official for the time being, considering that right now there aren't really any.

**Comment by DeeDeeG** [comment]
A rather late update, but this issue should finally be more neatly resolved "out of the box" (no `brew install libiconv` needed) by https://github.com/pulsar-edit/superstring/pull/15 and subsequently https://github.com/pulsar-edit/pulsar/pull/1051.
